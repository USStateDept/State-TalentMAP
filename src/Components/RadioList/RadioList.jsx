import { Component } from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import { EMPTY_FUNCTION } from '../../Constants/PropTypes';

class RadioList extends Component {
  constructor(props) {
    super(props);
    const value = get(props, 'options[0].value') || get(props, 'options[0].id');
    this.state = {
      value: props.value || value || '',
    };
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({ value: nextProps.value });
    }
  }

  updateValue = e => {
    const { onChange } = this.props;
    const value = get(e, 'target.value', null);
    if (value) {
      this.setState({ value }, () => {
        onChange(value);
      });
    }
  };

  render() {
    const { value: selectedValue } = this.state;
    const { options, ulProps, liProps, inputProps, labelProps } = this.props;
    return (
      <ul className="usa-unstyled-list" {...ulProps}>
        {
          options.map((o) => {
            const id = o.id;
            const name = o.name || id;
            const value = o.value || id;
            const label = o.label || id;
            const isChecked = selectedValue === value;
            return (
              <li key={id} {...liProps}>
                <input
                  onChange={this.updateValue}
                  id={id}
                  type="radio"
                  checked={isChecked}
                  name={name}
                  value={value}
                  {...inputProps}
                />
                <label htmlFor={id} {...labelProps}>{label}</label>
              </li>
            );
          })
        }
      </ul>
    );
  }
}

RadioList.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string,
      value: PropTypes.string,
      label: PropTypes.string,
    }),
  ).isRequired,
  value: PropTypes.string,
  onChange: PropTypes.func,
  ulProps: PropTypes.shape({}),
  liProps: PropTypes.shape({}),
  inputProps: PropTypes.shape({}),
  labelProps: PropTypes.shape({}),
};

RadioList.defaultProps = {
  value: '',
  onChange: EMPTY_FUNCTION,
  ulProps: {},
  liProps: {},
  inputProps: {},
  labelProps: {},
};

export default RadioList;
