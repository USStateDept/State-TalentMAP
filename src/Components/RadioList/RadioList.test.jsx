import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import RadioList from './RadioList';

describe('RadioListComponent', () => {
  const props = {
    options: [
      {
        id: 'a',
        name: 'a-name',
        value: 'a-value',
        label: 'a-label',
      },
      {
        id: 'b',
        name: 'b-name',
        value: 'b-value',
        label: 'b-label',
      },
    ],
    value: 'b-value',
  };

  it('is defined', () => {
    const wrapper = shallow(
      <RadioList {...props} />,
    );
    expect(wrapper).toBeDefined();
  });

  it('receives props', () => {
    const wrapper = shallow(
      <RadioList {...props} />,
    );
    expect(wrapper.instance().props.value).toEqual(props.value);
  });

  it('handles the onChange event', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <RadioList
        {...props}
        onChange={spy}
      />,
    );

    const input = wrapper.find('input').at(0);
    input.simulate('change', { target: { value: input.props().value } });
    expect(wrapper.instance().state.value).toBe(input.props().value);
    sinon.assert.calledOnce(spy);
  });

  it('sets new props', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <RadioList
        {...props}
        onChange={spy}
      />,
    );

    wrapper.setProps({ ...props, value: 'a-value' });
    expect(wrapper.instance().state.value).toBe('a-value');
  });

  it('matches snapshot', () => {
    const wrapper = shallow(
      <RadioList
        {...props}
      />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
