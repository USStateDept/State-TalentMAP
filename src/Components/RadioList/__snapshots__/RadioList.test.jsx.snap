// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioListComponent matches snapshot 1`] = `
<ul
  className="usa-unstyled-list"
>
  <li
    key="a"
  >
    <input
      checked={false}
      id="a"
      name="a-name"
      onChange={[Function]}
      type="radio"
      value="a-value"
    />
    <label
      htmlFor="a"
    >
      a-label
    </label>
  </li>
  <li
    key="b"
  >
    <input
      checked={true}
      id="b"
      name="b-name"
      onChange={[Function]}
      type="radio"
      value="b-value"
    />
    <label
      htmlFor="b"
    >
      b-label
    </label>
  </li>
</ul>
`;
